compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -403,23 +403,79 @@ LogicalResult getStaticDims(Operation *op,
   return success();
 }
 
+/// Utility to get the static offsets, sizes and strides for
+/// `AIEX::NpuDmaMemcpyNdOp` with implicit addressing.
+LogicalResult getStaticDimsForImplicitAddressing(
+    Operation *op, MemRefType memrefType, SmallVectorImpl<int64_t> &staticSizes,
+    SmallVectorImpl<int64_t> &staticStrides) {
+  SmallVector<int64_t> shapeArr;
+  for (auto shape : memrefType.getShape()) shapeArr.push_back(shape);
+  int64_t sizeIndex = staticSizes.size() - 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticSizes[sizeIndex--] = shapeArr[i];
+  }
+  int64_t strideIndex = staticStrides.size() - 1;
+  // Since aiex.npu.dma_cpy_nd expects 3-dimensional stride attribute, it has
+  // the 4-th implicit dimension as 1. Hence we store that in `prevStride`.
+  int64_t prevStride = 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticStrides[strideIndex] = shapeArr[i] * prevStride;
+    prevStride = staticStrides[strideIndex--];
+  }
+  for (int64_t i = strideIndex; i >= 0; i--) {
+    staticStrides[i] = staticStrides[i + 1];
+  }
+  return success();
+}
+
 /// Convert the `amdaie.npu.dma_cpy_nd` operation to `aiex.npu.dma_memcpy_nd`.
 LogicalResult npuDmaCpyNdOpToAIE(IRRewriter &rewriter,
                                  AMDAIE::NpuDmaCpyNdOp dmaOp,
                                  SmallVector<Operation *> &toBeErased,
                                  IRMapping &mapper, IRMapping &bindingsMapper) {
   rewriter.setInsertionPoint(dmaOp);
   // Convert bidirectional `amdaie.npu.dma_cpy_nd` op into two halves.
-  if (dmaOp.hasSourceAddressing()) {
+
+  // TODO(avarma): The logic for checking implicit L3 addressing should just be
```

Comment:
What do you mean with 'checking implicit L3 addressing'?

Reply:
So I was trying to refer to the following case :-
```
     %dma_target_l3 = amdaie.circular_dma_cpy_nd(%obj0[] [] [], %obj1[] [] []) : (!amdaie.logicalobjectfifo<memref<32x64xi32>>, !amdaie.logicalobjectfifo<memref<32x32xi32, 1>>)
    controlcode {
        %npu_dma_1 = amdaie.npu.dma_cpy_nd %dma_target_l3([] [] [], [] [] [])
        amdaie.npu.dma_wait(%npu_dma_1, S2MM)
    }
```

Where the access pattern is made implicit in `%npu_dma_1` and I'm trying to see through the `npu.dma_cpy_nd` source if the addressing on L3 is in source or target of `circular_dma_cpy_nd`.

Could you suggest a better way to word the same in comment?

Also, I wanted to confirm if the TODO's intention is valid - as in I wanted to have an API (defined in `AMDAIEOps.td`) within `npu.dma_cpy_nd` that'd answer if the op has implicit/explicit access/addressing patterns. If that's okay, I would like to add that within the same PR and use it in the appropriate place in the fix.

Reply:
> Also, I wanted to confirm if the TODO's intention is valid - as in I wanted to have an API (defined in AMDAIEOps.td) within npu.dma_cpy_nd that'd answer if the op has implicit/explicit access/addressing patterns. If that's okay, I would like to add that within the same PR and use it in the appropriate place in the fix.

What's the difference between that new api and the existing `hasSourceAddressing`?

Reply:
> Could you suggest a better way to word the same in comment?

I think the wording 'checking implicit L3 addressing' is ok! I am just not sure about what exact api you're thinking of?

Reply:
I think `hasImplicitSourceAddressing()` should just be `!hasExplicitSourceAddressing()`. If you also want to check for L3 access, I think it would be better to make that explicit in the function name and imo better to make that a separate function: `hasSourceOnL3` or maybe better `int64_t getSourceMemorySpaceInt()`

Reply:
Oh okay, cool.

Then I can add `getSourceMemorySpaceInt` and `getTargetMemorySpaceInt` instead for `npu.dma_cpy_nd` op and make use of those APIs here.

Reply:
Btw, for consistency, I used `uint64_t getMemorySpaceAsUInt()` before: https://github.com/nod-ai/iree-amd-aie/blob/741facf6ab616529978f077bd75b1a941b7385ac/compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td#L504. And also note that I used `0` for L3